src/poetry/installation/executor.py

```diff
@@ -603,34 +604,66 @@ def _prepare_archive(self, operation: Install | Update) -> Path:
 
         self._populate_hashes_dict(archive, package)
 
-        return self._chef.prepare(archive, editable=package.develop)
+        return self._chef.prepare(
+            archive, editable=package.develop, output_dir=output_dir
+        )
 
     def _prepare_git_archive(self, operation: Install | Update) -> Path:
         from poetry.vcs.git import Git
 
         package = operation.package
+        reference = package.source_resolved_reference or package.source_reference
+        assert package.source_url is not None
+        assert reference is not None
+
+        if package.source_resolved_reference is not None:
+            # Only cache git archives when we know precise reference hash,
+            # otherwise we might get stale archives
+            cached_archive = self._artifact_cache.get_cached_archive_for_git(
+                package.source_url,
+                package.source_resolved_reference,
+                package.source_subdirectory,
+                env=self._env,
+            )
+            if cached_archive is not None and not operation.package.develop:
+                return cached_archive
+
         operation_message = self.get_operation_message(operation)
 
         message = (
             f"  <fg=blue;options=bold>•</> {operation_message}: <info>Cloning...</info>"
         )
         self._write(operation, message)
 
-        assert package.source_url is not None
         source = Git.clone(
             url=package.source_url,
             source_root=self._env.path / "src",
-            revision=package.source_resolved_reference or package.source_reference,
+            revision=reference,
```

Comment:
We can revert this change (and storing reference in a variable some lines above) since we only use `source_resolved_reference` for caching now.